### PR TITLE
[no ticket][risk=no] Puppeteer test code refactor

### PR DESCRIPTION
**Pull request by aweng98**
`google-login.ts` page no longer extends `base-page.ts` to avoid running into circular dependency issue.

Organize some common functions hoping to make them easier to find:
* `BasePage.ts`'s `clickAndWait()` => `BaseElement.ts` `clickAndWait()`
* `app.ts`'s `signIn()` => `test-utils.ts` `signIn()`
* `authenticated-page.ts`'s `waitUntilNoSpinner()` => `test-utils.ts` `waitWhileLoading()`
* all `wait*` functions in `base-page.ts` => `waits-utils.ts`. so these functions can be used for pages that extends from `base-page.ts` as well as pages that do not extends from `base-page.ts` like `google-login.ts` page.
* `isLoaded()` function in every page has been refactored to show a user-friendly message when it fails, like `console.log(`HomePage isLoaded() encountered ${err}`);`.

Safely deleted unused classes:
* `label.ts`
* `css-selector-util.ts`
* `page-wait.ts`
* `xpath-util.ts`

### CODE REVIEW

File: e2e/jest.test-setup.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ beforeEach(async () => {
   await page.setUserAgent(userAgent);
   // Refer to https://github.com/puppeteer/puppeteer/blob/master/docs/api.md#pagesetdefaultnavigationtimeouttimeout
   await page.setDefaultNavigationTimeout(70000);
-  await page.setDefaultTimeout(10000);
+  await page.setDefaultTimeout(50000);
```
Comment (aweng98) [inline e2e/jest.test-setup.ts:10]: I changed it to 10s but it has been the source for couple failed runs today. Making timeOut bigger.

File: e2e/app/page/base-page.ts [hunk 2]
```diff
@@ -42,142 +41,11 @@ export default abstract class BasePage {
...
-
-    await this.page.waitForSelector(cssSelector, { visible: true });
-  }
-
   /**
    * Find texts in DOM.
    * @param txt
    */
-  async findText(txt: string): Promise<boolean> {
+  async containsText(txt: string): Promise<boolean> {
```
Comment (aweng98) [inline e2e/app/page/base-page.ts:48]: I think `containsText` is a better name. :)

File: e2e/app/page/create-account-page.ts [hunk 3]
```diff
@@ -155,14 +163,6 @@ export default class CreateAccountPage extends BasePage {
     await roleSelect.select(INSTITUTION_ROLE_VALUE.UNDERGRADUATE_STUDENT);
   }
 
-  // Step 2: Fill out Invitation key
-  async fillOutInvitationKey(invitationKey: string) {
-    await this.getInvitationKeyInput()
-    .then(invitationKeyInput => invitationKeyInput.type(invitationKey))
-    .then(() => this.getNextButton())
-    .then(submitButton => submitButton.click());
-  }
```
Comment (aweng98) [inline e2e/app/page/create-account-page.ts:164]: Unused because no more Invitation Key.

File: e2e/tests/workspace/workspace-clone.spec.ts [hunk 4]
```diff
@@ -1,11 +1,11 @@
 import WorkspacesPage from 'app/page/workspaces-page';
-import {signIn} from 'utils/app-utils';
+import {signIn} from 'utils/test-utils';
 import {WorkspaceAccessLevel, WorkspaceAction} from 'app/page-identifiers';
 import DataPage from 'app/page/data-page';
 import WorkspaceCard from 'app/component/workspace-card';
 import * as fp from 'lodash/fp';
 
-describe('Clone workspace', () => {
+describe.skip('Clone workspace', () => {
```
Comment (aweng98) [inline e2e/tests/workspace/workspace-clone.spec.ts:8]: un-skip before merge.

File: e2e/tests/workspace/workspace-create.spec.ts [hunk 5]
```diff
@@ -1,13 +1,12 @@
...
 import WorkspacesPage, {FIELD} from 'app/page/workspaces-page';
-import {signIn} from 'utils/app-utils';
+import {signIn, performActions} from 'utils/test-utils';
 import Button from 'app/element/button';
 import * as testData from 'resources/data/workspace-data';
-import {performActions} from 'utils/test-utils';
 import {makeWorkspaceName} from 'utils/str-utils';
 
-describe('Creating new workspaces', () => {
+describe.skip('Creating new workspaces', () => {
```
Comment (aweng98) [inline e2e/tests/workspace/workspace-create.spec.ts:9]: un-skip before merge.

File: e2e/app/page/authenticated-page.ts [hunk 6]
```diff
@@ -21,9 +20,13 @@ export default abstract class AuthenticatedPage extends BasePage {
   }
 
   protected async isSignedIn(): Promise<boolean> {
-    await this.page.waitForSelector(SELECTOR.signedInIndicator);
-    await this.page.waitForSelector(SELECTOR.logo, {visible: true});
-    return true;
+    try {
+      await this.page.waitForSelector(SELECTOR.signedInIndicator);
+      await this.page.waitForSelector(SELECTOR.logo, {visible: true});
```
Comment (NehaBroad) [inline e2e/app/page/authenticated-page.ts:25]: The AoU logo is visible on signed out/account creation pages as well

File: e2e/app/page/create-account-page.ts [hunk 7]
```diff
@@ -52,14 +54,16 @@ export default class CreateAccountPage extends BasePage {
...
-
-  async getInvitationKeyInput(): Promise<Textbox> {
-    const textbox = new Textbox(this.page);
-    await textbox.withCss('#invitationKey');
-    return textbox;
+    try {
+      await Promise.all([
+        waitForText(this.page, 'Please read through the entire agreement to continue'),
+        this.page.waitForXPath('//*[@data-test-id="account-creation-tos"]', {visible: true}),
+      ])
```
Comment (NehaBroad) [inline e2e/app/page/create-account-page.ts:61]: can you try this like:
Promise.all([...])
.then(values => { 
  return true
})
.catch(error => { 
  console.error(Message);return false
});

File: e2e/app/page/data-page.ts [hunk 8]
```diff
@@ -25,12 +27,13 @@ export default class DataPage extends AuthenticatedPage {
   async isLoaded(): Promise<boolean> {
     try {
       await Promise.all([
-        this.waitUntilTitleMatch(PAGE.TITLE),
+        await waitForDocumentTitle(this.page, PAGE.TITLE),
         this.page.waitForXPath(TAB_SELECTOR.showAllTab, {visible: true}),
-        this.waitUntilNoSpinner(),
+        waitWhileLoading(this.page),
       ]);
```
Comment (NehaBroad) [inline e2e/app/page/data-page.ts:33]: same here try if in formation Promise.all([...]).then(resolve).catch

File: e2e/app/page/create-account-page.ts (hunk 7 above)
Comment (NehaBroad) [reply e2e/app/page/create-account-page.ts:61]: eg  with await async function f() {

  let promise = new Promise((resolve, reject) => {
    setTimeout(() => resolve("done!"), 1000)
  });

  let result = await promise; // wait until the promise resolves (*)

  alert(result); // "done!"
}

File: e2e/app/page/workspace-edit-page.ts [hunk 9]
```diff
@@ -279,12 +281,16 @@ export default class WorkspaceEditPage extends AuthenticatedPage {
 
   async isLoaded(): Promise<boolean> {
     try {
-      await this.waitUntilTitleMatch(PAGE.TITLE);
-      await this.getWorkspaceNameTextbox();
-      await new SelectMenu(this.page, LABEL_ALIAS.SELECT_BILLING).getSelectedValue();
-      await this.getCreateWorkspaceButton();
+      await Promise.all([
+        waitForDocumentTitle(this.page, PAGE.TITLE),
```
Comment (NehaBroad) [inline e2e/app/page/workspace-edit-page.ts:285]: will it be ok if the request will happen in any order?

File: e2e/tests/workspace/workspace-create.spec.ts (hunk 5 above)
Comment (NehaBroad) [reply e2e/tests/workspace/workspace-create.spec.ts:9]: what do you mean?

File: e2e/app/page/authenticated-page.ts (hunk 6 above)
Comment (aweng98) [reply e2e/app/page/authenticated-page.ts]: you're correctly. I updated `signedInIndicator ` to be `app-signed-in`.

File: e2e/app/page/workspace-edit-page.ts (hunk 9 above)
Comment (aweng98) [reply e2e/app/page/workspace-edit-page.ts:285]: yes. what matters is that they're all present.

File: e2e/tests/workspace/workspace-create.spec.ts (hunk 5 above)
Comment (aweng98) [reply e2e/tests/workspace/workspace-create.spec.ts]: I skipped clone and create workspaces tests for the PR. I will undo skip first then merge to master.

**Pull request merged**